### PR TITLE
Fail pytest runs on warnings

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description
I've opened a few PRs over time to clean up warnings in the test suite when I see them. Figured we might as well fail CI so that we don't merge with known issues.

Failure expected due to #7795, mentioned here: https://github.com/PyCQA/pylint/pull/7795#issuecomment-1321240480. Edit: resolved.

### CODE REVIEW

File: tests/pyreverse/test_inspector.py [hunk 1]
```diff
@@ -30,7 +30,14 @@ def project(get_project: GetProjectCallable) -> Generator[Project, None, None]:
     with _test_cwd(TESTS):
         project = get_project("data", "data")
         linker = inspector.Linker(project)
-        linker.visit(project)
+        if "clientmodule_test" in str(project):
+            with pytest.warns(DeprecationWarning):
+                # tests.data.clientmodule_test.py emits DeprecationWarning
+                # due to use of __implements__
+                # TODO after deprecation: remove __implements__
```
Comment (Pierre-Sassoulas) [inline tests/pyreverse/test_inspector.py:37]:
```suggestion
                # TODO: 3.0 remove 'pytest.warns' once __implements__ is removed
```

**Review by clavedeluna — DISMISSED** [review]
Really good catch here and sorry one of my PRs introduced a warning. Looks like now it's failing the test run with it! Let me know if you want to fix the warnings in this PR or you want me to intervene via another issue/ PR.

**Comment by DudeNr33** [comment]
I have to look at this more closely this evening. Currently I don't see why this change should work as intended.

File: pyproject.toml [hunk 2]
```diff
@@ -89,6 +89,7 @@ test = "pytest"
 testpaths = ["tests"]
 python_files = ["*test_*.py"]
 addopts = "--strict-markers"
+filterwarnings = "error"
```
Comment (clavedeluna) [inline pyproject.toml:92]: will this make pytest fail at any len(warnings) > 0?

**Comment by jacobtylerwalls** [comment]
> Currently I don't see why this change should work as intended.

There's one pyreverse test that still makes use of `__implements__`, so it raises a DeprecationWarning when it's loaded. Or were you asking about something else?

File: pyproject.toml (hunk 2 above)
Comment (jacobtylerwalls) [reply pyproject.toml:92]: Yep!

**Comment by jacobtylerwalls** [comment]
> Let me know if you want to fix the warnings in this PR or you want me to intervene via another issue/ PR.

If you have time to help, that would be great. You can either push to my branch (GH's own editor makes that easier than it used to) or open a new PR, whatever works 👍🏻

File: tests/pyreverse/test_inspector.py [hunk 3]
```diff
@@ -30,7 +30,14 @@ def project(get_project: GetProjectCallable) -> Generator[Project, None, None]:
     with _test_cwd(TESTS):
         project = get_project("data", "data")
         linker = inspector.Linker(project)
-        linker.visit(project)
+        if "clientmodule_test" in str(project):
```
Comment (DudeNr33) [inline tests/pyreverse/test_inspector.py:33]: This is never true, as the `project` is always the same here (the `data` directory underneath `/tests`). The string representation of `project` does not contain the substring `clientmodule_test`.

I would suggest just get rid of the `if` condition here.

**Review by DudeNr33 — CHANGES_REQUESTED** [review]
Managed to squeeze the review in. As suspected the change does not help in this case, but I can understand the confusion (the `Generator` return type together with `TESTS` sounding like plural it seems like we are iterating over different directories here).

Let me know if you need more help!

File: tests/pyreverse/test_inspector.py (hunk 3 above)
Comment (jacobtylerwalls) [reply tests/pyreverse/test_inspector.py:33]: I tried that, but I received many failures. I'll try again tonight. 🤔

Comment (DudeNr33) [reply tests/pyreverse/test_inspector.py:33]: Hm, but with the current change we are always hitting the `else`, and nothing has changed effectively compared to before. 
I haven't looked at how to actually properly catch the warning, my guess would be that there are some other tests that don't catch this Warning yet (the majority of the legacy tests for `pyreverse` work with the same test data). I will also take a closer look at it this or tomorrow evening!

Comment (DudeNr33) [reply tests/pyreverse/test_inspector.py:33]: Neither on your branch nor on main could I reproduce the deprecation warnings on the `pyreverse` tests. 
On a [recent CI run](https://github.com/PyCQA/pylint/actions/runs/3514404778/jobs/5888375701) there is only one `DeprecationWarning` for some other test case, but not for `pyreverse`.
Am I missing something here like an option I have to turn on first?

Comment (jacobtylerwalls) [reply tests/pyreverse/test_inspector.py:33]: Sorry @DudeNr33 you're quite right, the `else` branch always executes. I've just pushed a change to just ignore warnings in this fixture.

**Review by Pierre-Sassoulas — APPROVED** [review]
Seems like no contributers used python 3.7 anymore :D

**Pull request merged**